### ISSUE TITLE
About classification evaluation

### ISSUE THREAD

**Issue by tomoki87** [masora1030/CVPR2023-FDSL-on-VisualAtom#3]
@masora1030 

Thank you for answering my question in another issue.
I would like to reproduce the classification task evaluation experiment described in the paper.
I am not sure how to acquire the data for Flowers, VOC12 and P30 and how to split the train/test.
Thank you in advance.

別issueでも私の疑問にご回答いただきありがとうございます。
論文記載の分類タスク評価の実験を再現したいと考えております。
Flowers,VOC12とP30のデータ取得方法とtrain/testのスプリット方法が分からないのですが、教えていただけますでしょうか？
よろしくお願いいたします。

![image](https://github.com/masora1030/CVPR2023-FDSL-on-VisualAtom/assets/106575909/e1291fcc-8040-430c-b186-904cf7900fce)

**Comment by tomoki87** [comment]
@masora1030 
I have an additional question. 
Regarding the models used in the evaluation described in the paper, 
could you also tell me whether the pretrain model or the finetune model is used for the last epoch or the best epoch? 
Thank you in advance.

追加で質問失礼いたします。
論文記載の評価で使用しているモデルに関しまして、
pretrainモデルとfinetuneモデルは、ラストエポックとベストエポックのどちらを使用させているかも教えていただけますでしょうか?
よろしくお願いいたします。

**Comment by masora1030** [comment]
@tomoki87 

We used the LAST epochs checkpoint of pre-training model and described the accuracy of the BEST epochs in the fine-tuning tasks.

pre-trainモデルはlast epochsのものを用い、fine-tuningのタスクはbest epochsの精度を記載しています。

**Comment by tomoki87** [comment]
@masora1030 

Thank you for your response.(I also appreciate your detailed answers to my other questions.)
I now understand the model used in the experiment.
I would appreciate it if you could also answer how you obtained the data for Flowers,VOC12 and P30 and how you split the train/test.
Thank you in advance.

ご回答いただきありがとうございます。(他の質問に対しても丁寧に回答いただき感謝します。)
実験で使用されたモデルについて理解できました。
Flowers,VOC12とP30のデータ取得方法とtrain/testのスプリット方法についてもご回答いただけますと幸いです。
よろしくお願いいたします。

**Comment by masora1030** [comment]
@tomoki87 
Sorry for the delay.
I'll answer about each dataset.

Flowers :. 
- dataset URL: http://host.robots.ox.ac.uk/pascal/VOC/voc2012/index.html#devkit
- Random sampling of 20 images for each class was used as train data, and the rest were used as test data.

VOC12 : 
- dataset URL : http://host.robots.ox.ac.uk/pascal/VOC/voc2012/index.html#devkit
- Originally split into train and valid

Places30 : dataset URL : 
- dataset URL : http://places2.csail.mit.edu/download.html
- class : Based on supplementary material in "Pre-training without Natural Images", Kataoka+, ACCV20 ( https://hirokatsukataoka.net/pdf/accv20_kataoka_fractaldb_supplementary.pdf ). Please see table 2.
- Random sampling of 5,000 images for each class was used as train data, and 100 images were used as test data.


お待たせして申し訳ございません。
それぞれのデータセットの準備について回答します。

Flowers : 
- dataset URL：http://host.robots.ox.ac.uk/pascal/VOC/voc2012/index.html#devkit
- 各クラスごとに20枚ずつランダムサンプリングしてtrain dataとし、残りはtest dataとした

VOC12 : 
- dataset URL : http://host.robots.ox.ac.uk/pascal/VOC/voc2012/index.html#devkit
- 元からtrainとvalidにスプリットされている

Places30 : 
- dataset URL : http://places2.csail.mit.edu/download.html
- class : "Pre-training without Natural Images", Kataoka+, ACCV20のサプリメンタリー資料https://hirokatsukataoka.net/pdf/accv20_kataoka_fractaldb_supplementary.pdf のtable 2を参照
- 各クラスごとに5000枚ずつランダムサンプリングしてtrain dataとし、100枚ずつをtest dataとした

**Comment by tomoki87** [comment]
@masora1030 
Thank you for your answer. 
I have taken your answer into consideration and have confirmed the accuracy reproduction of VOC12 and P30 in my environment.
We are experimenting with the following configuration in relation to Flowers, but the accuracy has deteriorated from 98.9 (value in the paper) to 92.1 (reproduction) in the model pre-trained with VisualAtom-1k.
- dataset：https://pytorch.org/vision/stable/generated/torchvision.datasets.Flowers102.html
- train and val in the above link are combined as training data (102 classes, 20 sheets for each class, 2040 sheets in total), and test (102 classes, irregular sheets for each class, 6149 sheets in total) is used for accuracy evaluation.

Since the batch size and lr are also aligned with reference to supplemental and can be reproduced with other evaluation data, we are wondering if there is a discrepancy in the composition of the Flowers data.
May I ask you to tell us about the following two points?
- The Flowers dataset URL that you provided in your response the other day seems to be the VOC one. Can you please share the URL again?
- If you have any points that you notice that are different between the above experimental conditions and the configuration used in the Flowers evaluation in the paper, please let us know.

Thank you in advance.

ご回答いただきありがとうございます。
いただいたご回答を参考にさせていただき、私の環境でもVOC12とP30の精度再現が確認できました。
Flowersに関しまして、以下の構成で実験しているのですが、VisualAtom-1kで事前学習したモデルにおいて、98.9(論文内の数値)→92.1(再現)と精度劣化している状態です。
- dataset：https://pytorch.org/vision/stable/generated/torchvision.datasets.Flowers102.html から取得
- 上記リンクのtrainとvalを合わせて学習データ(102クラス、各クラス20枚、計2040枚)とし、test(102クラス、各クラス不規則な枚数、計6149枚)を精度評価に使用

バッチサイズやlrもsupplementalを参考にして揃えていること、他評価データでは再現ができることからFlowersデータの構成に齟齬があるのではないかと考えております。
以下2点について教えていただいてもよろしいでしょうか？
- 先日ご回答いただいたFlowersのdataset URLがVOCのものになっていると思われます。再度URLを共有いただけますでしょうか？
- 上記の実験条件と論文のFlowers評価で使用した構成で異なる点に関して、お気づきの点があれば教えていただけますでしょうか？

よろしくお願いいたします。

**Comment by masora1030** [comment]
@tomoki87

> I have taken your answer into consideration and have confirmed the accuracy reproduction of VOC12 and P30 in my environment.

Great! Thank you for reproduction experiment.

Also, thank you for the additional question about Flowers dataset.
We will respond below.

> The Flowers dataset URL that you provided in your response the other day seems to be the VOC one. Can you please share the URL again?

Sorry, my mistake. the correct URL is  https://www.kaggle.com/datasets/nunenuh/pytorch-challange-flower-dataset .

> If you have any points that you notice that are different between the above experimental conditions and the configuration used in the Flowers evaluation in the paper, please let us know.

We have reconfirmed the composition of Flowers and found that the dataset path referred to in the previous confirmation was different, and we have provided you incorrect information about this. We deeply apologize for this.
At that time, we actually used the split of the dataset downloaded from the above URL as it was.
For fair accuracy comparisons, all pre-training methods compared in Table 7 used the same dataset and evaluation approach for Flowers.

> 私の環境でもVOC12とP30の精度再現が確認できました。

こちら、良かったです。再現実験を行っていただきありがとうございます。

また、追加でFlowersについてのご確認をいただきありがとうございます。
以下改めて回答いたします。

> 先日ご回答いただいたFlowersのdataset URLがVOCのものになっていると思われます。再度URLを共有いただけますでしょうか？

申し訳ございません。正しいURLは https://www.kaggle.com/datasets/nunenuh/pytorch-challange-flower-dataset です。

> 上記の実験条件と論文のFlowers評価で使用した構成で異なる点に関して、お気づきの点があれば教えていただけますでしょうか？

こちらについてFlowersのデータセット構成について再確認したところ、前回確認した際に参照したデータセットのpathが異なっており、誤った情報をお伝えしておりました。大変申し訳ございません。
当時実際に用いていたFlowersのsplitについては、上記URLからダウンロードしたデータセットのsplitをそのまま使用しておりました。
公正な精度比較のために、Table 7で比較されている全てのPre-Training methodにおいて、Flowersに対する評価方法は一貫して同様のデータセット、スプリットを使用しています。

以上よろしくお願いいたします。

**Comment by tomoki87** [comment]
@masora1030 
Thank you for your response.
I was able to get the data from the URL you shared.
I would like to ask two questions in reproducing the experiment.

- The accuracy reported in the paper is the value output by finetune.py.(https://github.com/masora1030/CVPR2023-FDSL-on-VisualAtom/blob/main/finetune.py#L276) In this case, which data in the dataset should be used for this calculation in Flowers, test or valid?
- If test is used in the above question, the data in the URL you shared with us does not seem to be divided into folders by class.
On the official Flowers website (https://www.robots.ox.ac.uk/~vgg/data/flowers/102/) , a list of image names and classes is available, and using this list, it was possible to sort the data into folders.Please let us know if you have taken any steps to address this issue, such as using a list.

Thank you in advance.


ご回答ありがとうございます。
共有いただいたURLからデータを取得することができました。
実験再現にあたって2点質問させていただきたいです。

- 論文で報告されている精度はfinetune.pyで出力される数値(https://github.com/masora1030/CVPR2023-FDSL-on-VisualAtom/blob/main/finetune.py#L276) を使用されているという認識です。この時、Flowersでこの計算を行うデータは、データセット内のtestとvalidどちらになりますでしょうか？

- 上の質問でtestを使用した場合、共有いただいたURLのデータではクラスごとにフォルダ分けをされていないようです。
公式のFlowersのサイト上(https://www.robots.ox.ac.uk/~vgg/data/flowers/102/) で画像名とクラスの対応リストが入手可能であり、このリストを使用すればフォルダ分けは可能でした。リストの使用等、何か対応をされていれば教えていただけますでしょうか？

よろしくお願いいたします。

**Comment by masora1030** [comment]
@tomoki87 

We apologize for the delay in replying and for the confusion caused by our poorly worded response.
We have used valid data when evaluated accuracy with Flowers in order to compare the accuracy under the same condition as in the previous work [28] (we have also confirmed with the author about the evaluation dataset). Also, as you pointed out, the test data is not directory-divided by category.
Therefore, we would appreciate it if you could use the valid data to reproduce or to compare the accuracy of our experiment with Flowers.
If you would like to compare our work with test data, we kindly ask you to conduct Fine-Tuning based on the pre-trained model with VisualAtom and then evaluate the results using the test data.

返信が遅くなってしまっていること、また、言葉足らずの回答でご混乱を招いてしまっており、大変申し訳ございません。
我々はFlowersを用いた全ての精度比較実験において、比較対象の先行研究[28]と同条件で精度比較を行うためにvalidデータを用いた評価を行っております（評価データセットについて著者に確認を行いました）。また、ご指摘の通りtestデータはカテゴリによってディレクトリ分けされておりません。
よって、我々のFlowersの実験精度を再現したり論文中の精度と比較する際には、validデータで評価いただければ幸いです。
testデータを用いて我々の研究成果との比較を行う場合には、お手数ですがVisualAtomでの事前学習済みモデルをベースにしたFine-Tuningをした上でtestデータによる評価を行っていただく形となります。

> [28] Kodai Nakashima, Hirokatsu Kataoka, Asato Matsumoto,
> Kenji Iwata, Nakamasa Inoue, and Yutaka Satoh. Can Vision Transformers Learn without Natural Images? In Proceedings of the AAAI Conference on Artificial Intelligence,
> volume 36, pages 1990–1998, 2022

**Comment by tomoki87** [comment]
@masora1030 
Sorry for the late reply.

I conducted a reproduction experiment with Flowers' vaild data,
I have confirmed that the accuracy is similar to that of the paper.

Thank you very much for answering my many questions about the classification task evaluation.
I have resolved the issue and will close the issue.

返信が遅くなり申し訳ございません。

Flowersのvaildデータで再現実験させていただいたところ、
論文と類似精度となることを確認しました。

分類タスク評価について多くの質問にご回答いただきありがとうございました。
本件解決いたしましたので、issueはcloseさせていただきます。
